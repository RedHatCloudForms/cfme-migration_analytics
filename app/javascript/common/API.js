const { API } = window;

export default {
  get(url, headers = {}, params = {}) {
    return API.get(url, {
      transformResponse: e => ({ data: e }),
      headers,
      params
    });
  },
  put(url, data = {}, headers = {}) {
    return API.put(url, data, {
      headers
    });
  },
  post(url, data = {}, headers = {}) {
    return API.post(url, data, {
      headers,
      transformResponse: e => ({ data: e })
    });
  },
  delete(url, headers = {}) {
    return API.delete(url, {
      headers
    });
  },
  patch(url, data = {}, headers = {}) {
    return API.patch(url, data, {
      headers
    });
  }
};
