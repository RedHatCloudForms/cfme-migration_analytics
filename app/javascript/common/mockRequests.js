let mockedRequests;

export const APImock = {
  reset: () => {
    mockedRequests = { GET: {}, POST: {}, PUT: {}, PATCH: {}, DELETE: {} };
  },
  mock: (method, url, status, response) => {
    mockedRequests[method][url] = { status, response };
  },
  respond: (method, url) => {
    const mockMatchingUrl = mockedRequests[method][url];
    const mocksMatchingMethod = Object.values(mockedRequests[method]);
    const mocked = mockMatchingUrl || (mocksMatchingMethod.length > 0 && mocksMatchingMethod[0]);
    if (mocked) {
      if (mocked.status === 200) {
        return Promise.resolve(mocked.response);
      }
      return Promise.reject(new Error('<mocked error>'));
    }
    return Promise.reject(new Error(`<no such mocked request ${method} ${url}>`));
  }
};

export const mockRequest = ({ method = 'GET', url, status = 200, response = null }) =>
  APImock.mock(method, url, status, response);

export const mockReset = () => APImock.reset();
