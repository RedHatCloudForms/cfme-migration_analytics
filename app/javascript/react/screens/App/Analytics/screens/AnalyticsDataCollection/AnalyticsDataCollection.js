import React from 'react';
import PropTypes from 'prop-types';
import { Spinner, Button } from 'patternfly-react';

const AnalyticsDataCollection = ({ onCancelClick }) => (
  <React.Fragment>
    <div className="large-spinner">
      <Spinner loading size="lg" inline />
      <h3>{__('Collecting inventory data')}</h3>
    </div>
    <Button onClick={onCancelClick}>{__('Cancel')}</Button>
  </React.Fragment>
);

AnalyticsDataCollection.propTypes = {
  onCancelClick: PropTypes.func.isRequired
};

export default AnalyticsDataCollection;
