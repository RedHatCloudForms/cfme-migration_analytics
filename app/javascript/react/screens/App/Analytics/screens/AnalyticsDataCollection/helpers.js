const payloadPathRegex = /\/migration_analytics\/.+$/;

export const getPayloadUrl = payloadPath => {
  const match = payloadPath && payloadPath.match(payloadPathRegex);
  if (!match) return null;
  const { protocol, hostname, port } = window.location;
  const origin = `${protocol}//${hostname}${port ? `:${port}` : ''}`;
  return `${origin}${match[0]}`;
};
