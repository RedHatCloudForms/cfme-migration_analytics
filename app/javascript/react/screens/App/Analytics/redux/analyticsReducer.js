import Immutable from 'seamless-immutable';
import {
  CALCULATE_SUMMARY_DATA,
  SELECT_PROVIDERS,
  SELECT_DETAILED_DATA,
  FETCH_MANIFEST_INFO,
  TOGGLE_MANIFEST_UPDATE_MODAL,
  START_INVENTORY_BUNDLE,
  FETCH_BUNDLE_TASK,
  RESET_DATA_COLLECTION_STATE,
  CHANGE_MANIFEST
} from './constants';
import { functionLookupReducer, getHandlersForBasicFetchActions } from '../../../../../redux/helpers';
import { calculateSummaryData } from './helpers';
import { RESET_ALL_STATE } from '../../../../../redux/common/constants';

const fetchManifestInfo = getHandlersForBasicFetchActions(
  FETCH_MANIFEST_INFO,
  'isFetchingManifestInfo',
  'errorFetchingManifestInfo',
  'manifestInfo'
);

const changeManifest = getHandlersForBasicFetchActions(
  CHANGE_MANIFEST,
  'isChangingManifest',
  'errorChangingManifest',
  'changeManifestResult'
);

const startInventoryBundle = getHandlersForBasicFetchActions(
  START_INVENTORY_BUNDLE,
  'isStartingBundle',
  'errorStartingBundle',
  'bundleStartResult'
);

const fetchBundleTask = getHandlersForBasicFetchActions(
  FETCH_BUNDLE_TASK,
  'isFetchingBundleTask',
  'errorFetchingBundleTask',
  'bundleTask'
);

export const initialState = Immutable({
  ...fetchManifestInfo.initialState,
  ...changeManifest.initialState,
  manifestUpdateModalVisible: false,
  summaryData: null,
  selectedProviders: [],
  detailedDataSelected: false,
  ...startInventoryBundle.initialState,
  ...fetchBundleTask.initialState
});

const actionHandlers = {
  ...fetchManifestInfo.actionHandlers,
  ...changeManifest.actionHandlers,
  [TOGGLE_MANIFEST_UPDATE_MODAL]: (state, action) =>
    state.set('manifestUpdateModalVisible', action.show !== null ? action.show : !state.manifestUpdateModalVisible),
  [CALCULATE_SUMMARY_DATA]: (state, action) => state.set('summaryData', calculateSummaryData(action.results)),
  [SELECT_PROVIDERS]: (state, action) => state.set('selectedProviders', action.selectedProviders),
  [SELECT_DETAILED_DATA]: (state, action) => state.set('detailedDataSelected', action.detailedDataSelected),
  ...startInventoryBundle.actionHandlers,
  ...fetchBundleTask.actionHandlers,
  [RESET_DATA_COLLECTION_STATE]: state =>
    Immutable.merge(state, {
      ...startInventoryBundle.initialState,
      ...fetchBundleTask.initialState
    }),
  [RESET_ALL_STATE]: () => initialState
};

export default functionLookupReducer(initialState, actionHandlers);
