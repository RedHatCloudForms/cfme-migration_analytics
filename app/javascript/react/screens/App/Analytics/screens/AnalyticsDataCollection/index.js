import { connect } from 'react-redux';
import AnalyticsDataCollection from './AnalyticsDataCollection';
import {
  startInventoryBundleAction,
  fetchBundleTaskAction,
  resetDataCollectionStateAction
} from '../../redux/analyticsActions';
import {
  selectIsPayloadReady,
  selectBundleError,
  selectBundleTaskHref,
  selectIsBundleTaskFinished,
  selectPayloadPath,
  selectPayloadHost,
  selectNumVms
} from '../../redux/analyticsSelectors';
import { getPayloadUrl } from './helpers';

const mapStateToProps = ({ migrationAnalytics: { analytics } }) => {
  const bundleError = selectBundleError(analytics);
  const isBundleTaskFinished = selectIsBundleTaskFinished(analytics);
  const payloadPath = selectPayloadPath(analytics);
  const payloadHost = selectPayloadHost(analytics);
  return {
    selectedProviders: analytics.selectedProviders,
    bundleError,
    bundleTaskHref: selectBundleTaskHref(analytics),
    isFetchingBundleTask: analytics.isFetchingBundleTask,
    isBundleTaskFinished,
    isPayloadReady: selectIsPayloadReady(analytics, { bundleError, isBundleTaskFinished }),
    numVms: selectNumVms(analytics),
    payloadHost,
    payloadPath,
    payloadUrl: getPayloadUrl(payloadPath)
  };
};

export default connect(
  mapStateToProps,
  { startInventoryBundleAction, fetchBundleTaskAction, resetDataCollectionStateAction }
)(AnalyticsDataCollection);
