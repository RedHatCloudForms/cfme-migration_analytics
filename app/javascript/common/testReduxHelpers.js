import React from 'react';
import thunkMiddleware from 'redux-thunk';
import promiseMiddleware from 'redux-promise-middleware';
import { createStore, applyMiddleware, combineReducers } from 'redux';
import configureMockStore from 'redux-mock-store';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';

const middlewares = [thunkMiddleware, promiseMiddleware];

export const generateStore = (reducers, initialState) =>
  createStore(combineReducers(reducers), initialState, applyMiddleware(...middlewares));

export const mockStore = initialState => configureMockStore(middlewares)(initialState);

export const mountWithTestStore = (reducers, initialState, children) => {
  const store = generateStore({ migrationAnalytics: combineReducers(reducers) }, { migrationAnalytics: initialState });
  return mount(<Provider store={store}>{children}</Provider>);
};

export const mockDispatch = actionCreator => {
  const dispatch = jest.fn();
  actionCreator(dispatch);
  return dispatch.mock.calls[0][0];
};
