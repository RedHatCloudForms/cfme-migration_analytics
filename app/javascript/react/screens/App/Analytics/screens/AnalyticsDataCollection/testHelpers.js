import { expectOnlySpecificActions } from '../../../../../../common/testReduxHelpers';

export const expectNoActionsOtherThan = (expectedActions, props) => {
  const allActions = ['startInventoryBundleAction', 'fetchBundleTaskAction', 'resetDataCollectionStateAction'];
  expectOnlySpecificActions({ allActions, expectedActions, props });
};
