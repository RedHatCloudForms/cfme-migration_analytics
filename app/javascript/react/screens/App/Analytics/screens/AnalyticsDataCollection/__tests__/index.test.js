import React from 'react';
import { noop } from 'patternfly-react';

import { mountWithTestStore } from '../../../../../../../common/testReduxHelpers';
import analyticsReducer from '../../../redux/analyticsReducer';

import AnalyticsDataCollection from '../AnalyticsDataCollection';
import AnalyticsDataCollectionContainer from '../index';

test('Data collection screen should mount with mapStateToProps reduced', () => {
  const wrapper = mountWithTestStore(
    { analytics: analyticsReducer },
    {},
    <AnalyticsDataCollectionContainer onCancelClick={noop} onReturnClick={noop} />
  );
  const component = wrapper.find(AnalyticsDataCollection);
  expect(component.props()).toMatchSnapshot();
});
