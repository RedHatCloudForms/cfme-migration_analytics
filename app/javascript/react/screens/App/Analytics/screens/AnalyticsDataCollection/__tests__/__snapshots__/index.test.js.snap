// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data collection screen should mount with mapStateToProps reduced 1`] = `
Object {
  "bundleError": null,
  "bundleTaskHref": null,
  "bundleTaskId": null,
  "downloadPayloadAction": [Function],
  "fetchBundleTaskAction": [Function],
  "isBundleTaskFinished": null,
  "isDownloadingPayload": false,
  "isFetchingBundleTask": false,
  "isPayloadReady": false,
  "numVms": null,
  "onCancelClick": [Function],
  "onReturnClick": [Function],
  "resetDataCollectionStateAction": [Function],
  "selectedProviders": Array [],
  "startInventoryBundleAction": [Function],
}
`;
